File: poirot/src/utils/system_info_reader.cpp
// Copyright 2026 Miguel Ángel González Santamarta
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "poirot/utils/system_info_reader.hpp"

#include <fstream>
#include <set>
#include <sys/sysinfo.h>
#include <sys/utsname.h>
#include <thread>

namespace poirot {
namespace utils {

CpuInfo SystemInfoReader::read_cpu_info() {
  CpuInfo cpu_info;
  cpu_info.cores = 0;

  std::ifstream cpuinfo("/proc/cpuinfo");
  if (cpuinfo.is_open()) {
    std::string line;
    std::set<int> physical_cores;
    int current_physical_id = -1;
    int current_core_id = -1;

    while (std::getline(cpuinfo, line)) {
      if (line.find("model name") != std::string::npos) {
        size_t pos = line.find(':');
        if (pos != std::string::npos) {
          cpu_info.model = line.substr(pos + 2);
        }

      } else if (line.find("core id") != std::string::npos) {
        size_t pos = line.find(':');
        if (pos != std::string::npos) {
          current_core_id = std::stoi(line.substr(pos + 2));
          // Combine physical_id and core_id to get unique physical cores
          int unique_core = (current_physical_id << 16) | current_core_id;
          physical_cores.insert(unique_core);
        }
      }
    }

    cpu_info.cores = physical_cores.empty()
                         ? static_cast<int>(std::thread::hardware_concurrency())
                         : static_cast<int>(physical_cores.size());
  }

  if (cpu_info.cores == 0) {
    cpu_info.cores = static_cast<int>(std::thread::hardware_concurrency());
  }

  return cpu_info;
}

MemoryInfo SystemInfoReader::read_memory_info() {
  MemoryInfo mem_info;
  mem_info.total_kb = 0;

  struct sysinfo info;
  if (sysinfo(&info) == 0) {
    mem_info.total_kb = static_cast<long>(info.totalram * info.mem_unit / 1024);
  }

  return mem_info;
}

OsInfo SystemInfoReader::read_os_info() {
  OsInfo os_info;

  struct utsname uts;
  if (uname(&uts) == 0) {
    os_info.name = uts.sysname;
    os_info.version = uts.release;
    os_info.hostname = uts.nodename;
  }

  std::ifstream os_release("/etc/os-release");
  if (os_release.is_open()) {
    std::string line;
    while (std::getline(os_release, line)) {
      if (line.find("PRETTY_NAME=") == 0) {
        os_info.name = line.substr(13);
        if (!os_info.name.empty() && os_info.name.back() == '"') {
          os_info.name.pop_back();
        }
        break;
      }
    }
  }

  return os_info;
}

} // namespace utils
} // namespace poirot

File: poirot/poirot/utils/system_info_reader.py
# Copyright 2026 Miguel Ángel González Santamarta
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import socket
import platform
from dataclasses import dataclass


@dataclass
class CpuInfo:
    """Structure to hold CPU information."""

    model: str = ""
    cores: int = 0


@dataclass
class MemoryInfo:
    """Structure to hold Memory information."""

    total_kb: int = 0


@dataclass
class OsInfo:
    """Structure to hold OS information."""

    name: str = ""
    version: str = ""
    hostname: str = ""


class SystemInfoReader:
    """
    Class for reading system information.

    Provides methods to read CPU, memory, and OS information.
    """

    def __init__(self) -> None:
        """Default constructor."""
        pass

    def read_cpu_info(self) -> CpuInfo:
        """
        Read CPU information from /proc/cpuinfo.

        Returns:
            CpuInfo struct containing model and core count.
        """
        cpu_info = CpuInfo()

        try:
            with open("/proc/cpuinfo", "r") as f:
                content = f.read()

            # Parse model name and count physical cores
            physical_cores = set()
            current_physical_id = -1
            current_core_id = -1

            for line in content.split("\n"):
                if line.startswith("model name"):
                    cpu_info.model = line.split(":")[1].strip()
                elif line.startswith("physical id"):
                    current_physical_id = int(line.split(":")[1].strip())
                elif line.startswith("core id"):
                    current_core_id = int(line.split(":")[1].strip())
                    # Combine physical_id and core_id to get unique physical cores
                    unique_core = (current_physical_id << 16) | current_core_id
                    physical_cores.add(unique_core)

            # Use physical core count if available, otherwise fall back to hardware_concurrency
            if physical_cores:
                cpu_info.cores = len(physical_cores)
            else:
                cpu_info.cores = os.cpu_count() or 1

        except OSError:
            cpu_info.model = platform.processor() or "Unknown"
            cpu_info.cores = os.cpu_count() or 1

        if cpu_info.cores == 0:
            cpu_info.cores = os.cpu_count() or 1

        return cpu_info

    def read_memory_info(self) -> MemoryInfo:
        """
        Read total system memory.

        Returns:
            MemoryInfo struct containing total memory in KB.
        """
        mem_info = MemoryInfo()

        try:
            with open("/proc/meminfo", "r") as f:
                for line in f:
                    if line.startswith("MemTotal:"):
                        # Parse value like "MemTotal:       16384000 kB"
                        parts = line.split()
                        if len(parts) >= 2:
                            mem_info.total_kb = int(parts[1])
                        break
        except (OSError, ValueError):
            pass

        return mem_info

    def read_os_info(self) -> OsInfo:
        """
        Read OS information from uname and /etc/os-release.

        Returns:
            OsInfo struct containing OS name, version, and hostname.
        """
        os_info = OsInfo()

        # Get uname info first (like C++ does)
        os_info.name = platform.system()  # sysname from uname
        os_info.version = platform.release()  # release from uname
        os_info.hostname = socket.gethostname()  # nodename from uname

        # Try to read PRETTY_NAME from /etc/os-release (overrides uname sysname)
        try:
            with open("/etc/os-release", "r") as f:
                for line in f:
                    if line.startswith("PRETTY_NAME="):
                        # Extract value after PRETTY_NAME="
                        value = line[13:].strip()
                        if value.endswith('"'):
                            value = value[:-1]
                        os_info.name = value
                        break
        except OSError:
            pass

        return os_info
